.github/workflows/python.yml

```diff
@@ -171,6 +171,23 @@ jobs:
         with:
           fetch-depth: 0
           submodules: recursive
+      - name: Install ccache
```

Comment:
hello @kou

Since I was interested in this issue, I added the following code based on other code.

In Ubuntu, I understood that the cache was applied by recycling the image.

However, despite using the cache, the overall CI speed does not seem to have improved much. And in case of macos 12, you can see that the cache hit rate result of ccache is 0.

Does this fit the intention?

#### AMD64 macOS 12 Python 3

https://github.com/apache/arrow/actions/runs/8831660979/job/24247320689?pr=41378#step:11:42

#### ARM64 macOS 14 Python 3

https://github.com/apache/arrow/actions/runs/8831660979/job/24247321146?pr=41378#step:11:42

Reply:
Hey! So the issue here is that `actions/cache` doesn't really work great for caching build caches like ccache as caches are immutable and if a primary key was hit it will not update the cache, leading to worse and worse hit rates.

In this case you also used the same key for 2 different jobs, this will lead to a key clash and prevent one job from caching (looking at the results likely arm64 was cached -> 0% hit rate on amd64 as ccache is os/arch specific).

A workaround is to create unique keys (e.g. `ccache-python-macos-<arch>-${{ github.sha }}` on each run and use 'restore-keys: ccache-python-macos-<arch>-' to still get a hit. This is quite wasteful and leads to frequent cache evictions as only 10GB/repo are allowed which we frequently go over (massivly)
![image](https://github.com/apache/arrow/assets/16141871/8cf7ccb4-8e41-4d60-a74f-d23bb3810d14)

I created an [alternative](https://github.com/assignUser/stash) for this case but have to ask infra to allow-list it.

Reply:
Thank you for the review.

How long will it take for an alternative solution to be allowed?

As you said, we understood that the cache was not configured properly because the same key was used on both platforms, and we fixed this.

#### AMD64 macOS 12 Python 3

https://github.com/apache/arrow/actions/runs/8858657235/job/24327563485?pr=41378#step:11:42

#### ARM64 macOS 14 Python 3

https://github.com/apache/arrow/actions/runs/8858657235/job/24327563485?pr=41378#step:11:42

Furthermore, we understand that the cache capacity is fundamentally limited and the cache is removed frequently, making it not used efficiently. If what you mentioned is applied, I expect that overall CI time will be reduced.

And until an alternative solution is applied, would it be better to proceed with the PR as a draft?